server/jetstream_batching.go

```diff
@@ -41,26 +42,71 @@ type batchGroup struct {
 	timer *time.Timer
 }
 
-func newBatchGroup(s *Server, batchId string, batches *batching, store StreamStore) *batchGroup {
-	b := &batchGroup{store: store}
+// Lock should be held.
+func (batches *batching) newBatchGroup(mset *stream, batchId string) (b *batchGroup, err error) {
+	store, err := newBatchStore(mset, batchId)
+	if err != nil {
+		return nil, err
+	}
+	b = &batchGroup{store: store}
 
 	// Create a timer to clean up after timeout.
 	timeout := streamMaxBatchTimeout
-	if maxBatchTimeout := s.getOpts().JetStreamLimits.MaxBatchTimeout; maxBatchTimeout > 0 {
+	if maxBatchTimeout := mset.srv.getOpts().JetStreamLimits.MaxBatchTimeout; maxBatchTimeout > 0 {
 		timeout = maxBatchTimeout
 	}
 	b.timer = time.AfterFunc(timeout, func() {
 		b.cleanup(batchId, batches)
 	})
-	return b
+	return b, nil
 }
 
-// stopCleanupTimerLocked indicates the batch is ready to be committed.
+func getBatchStoreDir(mset *stream, batchId string) (string, string) {
+	mset.mu.RLock()
+	jsa, name := mset.jsa, mset.cfg.Name
+	mset.mu.RUnlock()
+
+	jsa.mu.RLock()
+	sd := jsa.storeDir
+	jsa.mu.RUnlock()
+
+	bname := getHash(batchId)
+	return bname, filepath.Join(sd, streamsDir, name, batchesDir, bname)
+}
+
+func newBatchStore(mset *stream, batchId string) (store StreamStore, err error) {
+	mset.mu.RLock()
+	replicas, storage := mset.cfg.Replicas, mset.cfg.Storage
+	mset.mu.RUnlock()
+
+	if replicas == 1 && storage == FileStorage {
+		bname, storeDir := getBatchStoreDir(mset, batchId)
+		fcfg := FileStoreConfig{AsyncFlush: true, BlockSize: defaultLargeBlockSize, StoreDir: storeDir}
+		s := mset.srv
+		prf := s.jsKeyGen(s.getOpts().JetStreamKey, mset.acc.Name)
+		if prf != nil {
+			// We are encrypted here, fill in correct cipher selection.
+			fcfg.Cipher = s.getOpts().JetStreamCipher
+		}
+		oldprf := s.jsKeyGen(s.getOpts().JetStreamOldKey, mset.acc.Name)
+		cfg := StreamConfig{Name: bname, Storage: FileStorage}
+		store, err = newFileStoreWithCreated(fcfg, cfg, time.Time{}, prf, oldprf)
+	} else {
+		store, err = newMemStore(&StreamConfig{Name: _EMPTY_, Storage: MemoryStorage})
+	}
+	return store, err
```

Comment:
In this case both `store` and `err` are used in either outcome, so you can just do `return` instead of `return store, err`.

Reply:
Done